File: Leetcode/Leetcode-DSA-course/Linked-List/82-remove-duplicates-from-sorted-list-ii/remove-duplicates-from-sorted-list-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* deleteDuplicates(ListNode* head) {
        if (head == nullptr || head->next == nullptr)
            return head;
        ListNode *newHead = nullptr, *end = nullptr, *curr = head;
        int prev = -111;
        while (curr->next != nullptr) {
            if (prev != curr->val && curr->val != curr->next->val) {
                if (newHead == nullptr) {
                    newHead = curr;
                    end = curr;
                } else {
                    end->next = curr;
                    end = curr;
                }
            } else {
                prev = curr->val;
            }
            curr = curr->next;
        }
        if (curr->val != prev) {
            if (end)
                end->next = curr;
            else newHead=curr;
        } else {
            if (end)
                end->next = nullptr;
        }
        return newHead;
    }
};

File: Leetcode/Leetcode-DSA-course/Linked-List/82-remove-duplicates-from-sorted-list-ii/remove-duplicates-from-sorted-list-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode deleteDuplicates(ListNode head) {
        // Sentinel
        ListNode sentinel = new ListNode(0, head);

        // predecessor = the last node 
        // before the sublist of duplicates
        ListNode pred = sentinel;
        
        while (head != null) {
            // If it's a beginning of the duplicates sublist 
            // skip all duplicates
            if (head.next != null && head.val == head.next.val) {
                // Move till the end of the duplicates sublist
                while (head.next != null && head.val == head.next.val) {
                    head = head.next;    
                }

                // Skip all duplicates
                pred.next = head.next;
    
            // otherwise, move predecessor
            } else {
                pred = pred.next;    
            }
                
            // move forward
            head = head.next;    
        }  
        return sentinel.next;
    }
}